scrunch/order.py

```diff
@@ -0,0 +1,576 @@
+                        obj.name: _get(obj)
+                    })
+                else:
+                    if isinstance(obj, scrunch.datasets.Variable):
+                        _id = obj.id
+                    else:
+                        _id = obj['id']
+                    _elements.append('../%s/' % _id)
+            return _elements
+
+        return _get(self.graph)
+
+    def update(self):
+        updated_order = {
+            'element': 'shoji:order',
+            'graph': self._build_graph_structure()
+        }
+        try:
+            self.hier.put(updated_order)
+        except (pycrunch.ClientError, pycrunch.ServerError) as e:
+            # Our update to the Hierarchical Order failed. Better reload.
+            self._hier = self._vars = self._graph = None
+            self._load_graph()
+            raise OrderUpdateError(str(e))
+
+    # Proxy methods for the __root__ Group
+
+    def __str__(self):
+        return str(self.graph)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __iter__(self):
+        return self.graph.itervalues()
+
+    def itervalues(self):
+        return self.graph.itervalues()
+
+    def iterkeys(self):
+        return self.graph.iterkeys()
+
+    def keys(self):
+        return self.graph.keys()
+
+    def values(self):
+        return self.graph.values()
+
+    def items(self):
+        return self.graph.items()
+
+    def __getitem__(self, item):
+        return self.graph[item]
+
+
+class DatasetVariablesOrder(Order):
+    def __init__(self, resource):
+        super(DatasetVariablesOrder, self).__init__(resource)
+        self._vars = None
+        self._revision = None
+
+    def _load_hier(self):
+        self._hier = self.resource.session.get(
+            self.resource.variables.orders.hier).payload
+        return self._hier
+
+    def _load_vars(self):
+        self._vars = self.resource.variables.by('id')
+        return self._vars
+
+    @property
+    def vars(self):
+        if self._vars is None:
+            self._load_vars()
+        return self._vars
+
+    @vars.setter
+    def vars(self, _):
+        raise TypeError('Unsupported assignment operation')
+
+    def get(self):
+        # Returns the synchronized hierarchical order graph.
+        if self._sync:
+            ds_state = self.resource.session.get(
+                self.resource.state.self).payload
+            if self._revision is None:
+                self._revision = ds_state.body.revision
+            elif self._revision != ds_state.body.revision:
+                # There's a new dataset revision. Reload the
+                # hierarchical order.
+                self._revision = ds_state.body.revision
+                self._load_hier()
+                self._load_graph()
+        return self
+
+
+class ProjectDatasetsOrder(Order):
+    def __init__(self, resource):
+        super(ProjectDatasetsOrder, self).__init__(resource)
+        self._datasets = None
+
+    def _load_hier(self):
```

Comment:
Almost all the subclass methods here can disappear. Just have the subclass be like this:
```
def __init__(self, order, catalog):
    super()...
    self.datasets = catalog.by('id')
```

And you won't need any of the bottom resources.